mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp

```diff
@@ -1000,6 +1000,26 @@ convertOmpWsLoop(Operation &opInst, llvm::IRBuilderBase &builder,
   return success();
 }
 
+/// Replace the region arguments of the parallel op (which correspond to private
+/// variables) with the actual private varibles they correspond to. This
+/// prepares the parallel op so that it matches what is expected by the
+/// OMPIRBuilder.
+static void prepareOmpParallelForPrivatization(omp::ParallelOp opInst) {
```

Comment:
Could you elaborate on why this is necessary to match what the OMPIRBuilder expects? My understanding is that this would make the following transformation:

```mlir
// Before.
%x = ...
omp.parallel private(@x.privatizer %x -> %arg0 : !llvm.ptr) {
^bb0(%arg0: !llvm.ptr):
  %0 = llvm.load %arg0 : !llvm.ptr -> i32
  omp.terminator
}

// After.
%x = ...
omp.parallel private(@x.privatizer %x -> %??? : !llvm.ptr) {
^bb0():
  %0 = llvm.load %x : !llvm.ptr -> i32
  omp.terminator
}
```
If that's the case, it also seems like it is making the `omp.parallel` operation invalid, since the number of block arguments wouldn't match the number of private variables. Is there a chance that the `bodyGenCB` callback could be tweaked to address the problems that this function solves by making changes to the original operation?

Maybe I'm just not understanding how this function works, so some clarification in that case would be much appreciated.

Reply:
> Could you elaborate on why this is necessary to match what the OMPIRBuilder expects?

The `OpenMPIRBuilder::createParallel(...)` [collects the inputs and outputs of the parallel region](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp#L1502). The [inputs are then iterated](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp#L1580) to invoke the private/firstprivate/shared callback.

The input as defined by the `CodeExtractor` is [any value defined outside the region and used inside it](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Utils/CodeExtractor.cpp#L649).

Therefore, if `OpenMPIRBuilder::createParallel` is invoked with the block arguments still used inside the parallel region, the IR builder won't detect these as inputs and won't invoke the `PrivCB` for private/firstprivate variables.

That's why `prepareOmpParallelForPrivatization` "rewires" the uses of private vars inside the region from the block args to the original SSA values.

> it also seems like it is making the omp.parallel operation invalid

Indeed, this does invalidate the op. But my understanding is that the op is at the end of its life at this stage and there is no need to keep it valid. However, to make this transformation less damaging, I removed the line that erases the region arguments. That way the region arguments are still there but just not used anymore. The op then is at a transitional state where original SSA values for private variables are used inside the region but the privatization logic is not inlined yet.

Reply:
Thank you for the explanation, I understand the problem a bit better now. At the moment it might be the case that this is the last time the `omp.parallel` operation is used, but in general we shouldn't make modifications to the MLIR operations during translation to LLVM IR or we would introduce some hard to debug problems in the future.

When processing the `omp.private` ops you're creating clones, making changes to those and then deleting them, could something like this be done for `omp.parallel`? That way we wouldn't have to rely on the `omp.parallel` operation never being used again after translation.

Reply:
Done. The per-processing function now returns a clone of the original op and the clone is used for code-gen and then deleted to clean up the module afterwards.

Reply:
> The OpenMPIRBuilder::createParallel(...) [collects the inputs and outputs of the parallel region](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp#L1502). The [inputs are then iterated](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp#L1580) to invoke the private/firstprivate/shared callback.
> The input as defined by the CodeExtractor is [any value defined outside the region and used inside it](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Utils/CodeExtractor.cpp#L649).
> Therefore, if OpenMPIRBuilder::createParallel is invoked with the block arguments still used inside the parallel region, the IR builder won't detect these as inputs and won't invoke the PrivCB for private/firstprivate variables.

Wouldn't it still be called for the original variable (`%x` here)? And at that point can't we get (`%arg0`) and do the appropriate processing?

Reply:
> Wouldn't it still be called for the original variable (%x here)? And at that point can't we get (%arg0) and do the appropriate processing?

No. To illustrate the difference, I created 2 commits (both will be reverted but just to clarify what is happening):

1. 363ae7f6afdd61f6f350eef7c38d325a0bec7170 where I still pre-proccess the op and also added logs to `OMPIRBuilder` to list the collected inputs. Here is the list of collected inputs:
```
>>>> collected input:   %tid.addr = alloca i32, align 4
>>>> collected input:   %zero.addr = alloca i32, align 4
>>>> collected input: ptr %0
>>>> calling privCB for: ptr %0
```

2. 3933d26d1a8a3cb645d6e82849e3d3bb5a4c2936 where I do not pre-proccess the op with the same logs as the above:
```
>>>> collected input:   %tid.addr = alloca i32, align 4
>>>> collected input:   %zero.addr = alloca i32, align 4
```

So the private variable is not detected as input if we don't do the pre-processing. And this actually matches my understanding of how the `CodeExtractor` works [when it collect inputs/outputs to the region](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Utils/CodeExtractor.cpp#L649).

I think we can modify the `CodeExtractor` to detect region operands as inputs but I think there will be other problems down the line.

Let me know if I missed anything.

(both test commits will be reverted before merging of course)

Reply:
@kiranchandramohan please have a look at the above comment and let me know if you have any concerns or comments 🙏.

Reply:
>  and add some findPrivVars that checks for if(definedInCaller(Blocks, V)) and returns the variables?

I think that will not work. If you check the definition of [definedInCaller](https://github.com/llvm/llvm-project/blob/f740366fa68d3cfceda7efe2d573348253fbb1e9/llvm/lib/Transforms/Utils/CodeExtractor.cpp#L279), you will see that it returns `true` only if a value is defined by an instruction that belongs to a block that is not an element in the `Blocks` argument to the function.

So, if you do not use `findInputsOutputs` and instead directly create a `findPrivVars` that directly calls `if(definedInCaller(Block, V))` without iterating over all instructions in the outlined region as `findInputsOutputs` [does](https://github.com/llvm/llvm-project/blob/f740366fa68d3cfceda7efe2d573348253fbb1e9/llvm/lib/Transforms/Utils/CodeExtractor.cpp#L652), I think you will falsely detect too many values as being private.

Please let me know if I misunderstood your suggestion! :)

Also, what is the disadvantage of the current approach of pre-processing the `omp.parallel` op by remapping the block argument uses to the original SSA values that will be privatized?

Reply:
> Also, what is the disadvantage of the current approach of pre-processing the `omp.parallel` op clone by remapping the block argument uses to the original SSA values that will be privatized?

Thanks for the analysis. I'm okay with the parallel clone op, if there are no alternatives. Let us wait for others to comment though; I am not super confident on this.